docs/lss.man
.\" Process this file with
.\" groff -man -Tascii lss.man
.\"
.TH LSS 1 "2012" OpenLSS "Website Package Manager"
.SH NAME
lss \- The OpenLSS package manager
.SH SYNOPSIS
lss [OPTIONS] --update

lss [OPTIONS] [--install --remove --purge] <packages>

lss [OPTIONS] [--show --set --add --del] --name <name> --value <value>

lss [OPTIONS] [UTILITY]

lss [OPTIONS] [LOCAL DB]

lss [OPTIONS] [DEV]

lss [OPTIONS] [PACKAGE DEV]

.SH DESCRIPTION
.B lss
manages the local package database for a target that is
running the OpenLSS framework.
.SH OPTIONS
.IP "-h --help"
Display help information
.IP "-y --yes"
Answer yes to all confirmation prompts
.IP "-v -vv -vvv"
Increase verbosity and show underlying system commands:
0 for standard, 1 for verbose, 2 for info, 3 for debug
.IP "-q -qq -qqq"
Decrase verbosity and suppress errors, warnings etc.
Use more to negate -v arguments
.SH RUN-TIME SETTINGS
.IP "-t --target <tgt>"
Manually specify the target (this should be set into the respective def)
.IP "-m --mirror <mirror>"
Specify external mirror (mainly for dev functions)
.SH PACKAGE MANAGEMENT
.IP "-U --update"
Syncs the local package database with the upstream
.IP "-s --search <keywords>"
Search the package database
.IP "-u --upgrade"
Upgrade the existing target
.IP "-i --install <packages>"
Install packages, can be a CSV of packages.
.IP "-r --remove <packages>"
Remove packages, can be a CSV of packages.
.IP "-p --purge <packages>"
Purge packages, like remove but processes additional removal hooks. Can be a CSV.
.SH LOCAL DB
.IP "--show-db"
Shows the local package db.
.IP "--db-file <file>"
When used with --show-db will show contents of a non standard package DB
.IP "-l --list"
List locally installed packages
.SH UTILITY
.IP "--int-version <version>"
Show the integer version of a standard version
.IP "--clear-cache"
Clear the local cache (must update db afterwards)
.IP "--version -V"
Show the current versions of the software
.SH DEV
.IP "-b --build-db"
Builds the package database from a local tree.
.IP "--export-db --mirror=<mirror>"
Export the package database to the remote mirror. (requires --mirror)
.SH SETTINGS
All of these commands take %fqn% which is a Fully Qualified Name of a package such as "main/util/func"
.IP "--show=::<sys|usr|tgt|%fqn%>"
Show a respective definition file
.IP "--set=::<sys|usr|tgt|%fqn%>"
Set a setting value for the respective def (usr|tgt options, defaults to sys)
.IP "--add=::<sys|usr|tgt|%fqn%>"
Add a value to an array for a respective def
.IP "--del=::<sys|usr|tgt|%fqn%>"
Removes a value from an array for a respective def
.IP "--name <name>"
Name of the value to be modified
.IP "--value <value>"
Value to be set
.SH PACKAGE DEV
.IP "--export=<%fqn%> --mirror=<mirror>"
Export a package to the respective mirror
.IP "--refactor=<%fqn%> --dir=<dir>"
Pull files from a working directory to package based on the manifest
.IP "--create=<%fqn%>"
Create the basic def file for a package and package contents folder
.IP "--delete=<%fqn%>"
Delete the def file of a package, remaining files must be removed manually
.SH EXAMPLES
.IP "Update Package Database"
lss --update
.IP "Install Packages"
lss --install web,admin,staff
.IP "Remove Package"
lss --remove web
.IP "Purge Packages"
lss --purge staff,admin
.IP "Add Mirror to User Def"
lss --add=user --name mirror --value http://mirror.openlss.org
.IP "Add Restricted Mirror to Target Def"
lss --add=target --name mirror --value http://username:password@mirror.openlss.org/private
.IP "Change Default Target"
lss --set --name target --value /opt/lss
.IP "Change Default UI to Menu"
lss --set --name ui --value 1
.IP "Update package information"
lss --set=main/util/func --name info.version --value 0.1.0
.IP "Export a package"
lss --export=main/util/func --mirror=/data/mirror
.SH BUGS
Report bugs at http://bugs.openlss.org
.SH DOCUMENTATION
See more documentation at http://wiki.openlss.org
.SH SUPPORT
Visit our community forum http://forum.openlss.org

See our mailing list at http://list.openlss.org

For commercial support see http://support.openlss.com
.SH AUTHORS
Bryan Tong <contact [at] nullivex.com>

Tony Butler <spudz76 [at] gmail.com>
.SH COPYRIGHT
(c) 2012 EggPire LLC, All Rights Reserved.

.SH LICENSE
LGPLv3+: GNU LGPL version 3 or later <http://www.gnu.org/licenses/lgpl.html>.

This is free software: you are free to change and redistribute it. There is NO WARRANTY, to the extent permitted by law.


core/CHANGELOG
0.1.1
-----------------------
* Fixed email validation to be more compliant

0.1.0
-----------------------
* Initial relese


rls/deb/lss-0.1.2/usr/lss/tools/lss.php
#!/usr/bin/php
<?php

require('src/boot.php');

//require sources
require_once(ROOT.'/tools/lib/pkg.php');
require_once(ROOT.'/tools/lib/pkgdef.php');
require_once(ROOT.'/tools/lib/pkgdb.php');
require_once(ROOT.'/tools/lib/pkgexport.php');
require_once(ROOT.'/tools/lib/tgtdef.php');
require_once(ROOT.'/tools/lib/db.php');

//control funcs
require_once(ROOT.'/tools/src/func_lss.php');

//set this to throw an error if there is no action
$noerror = false;

//figure our opts
$so = array(
	//command settings
	'h', //help
	'y', //yes

	//package actions
	'u', //upgrade
	's:', //search
	'i:', //install
	'r:', //remove
	'p:', //purge
	'l', //list installed

	//backup/restore
	'B::', //backup
	'R::', //restore

	//db actions
	'U', //update
	'b', //build-db
	'e', //export-db
	'S', //show-db
	
	//misc
	'V', //version

	//working dir
	't:', //user working directory

	//mirrors
	'm:', //use mirror
);
$lo = array(
	//command settings
	'help',
	'yes',

	//package actions
	'upgrade',
	'search:',
	'install:',
	'remove:',
	'purge:',
	'list',

	//backup/restore
	'backup::',
	'restore::',
	'restore-file:',
	'migrate:',
	'backup-list',
	'db-dump:',
	'db-restore:',

	//db actions
	'db-file:',
	'build-db',
	'export-db',
	'show-db',
	'update',

	//mirror
	'mirror:',

	//target
	'target:',
	
	//working dir
	'dir:',

	//misc/utility
	'int-version:',
	'clear-cache',
	'version',

	//def management options
	'set::',
	'add::',
	'del::',
	'show::',
	'name:',
	'value:',
	
	//pkg specific
	'create:',
	'delete:',
	'export:',
	'refactor:',

);
$opts = getopt(implode('',$so),$lo); unset($so,$lo);

//figure out our answer status
if(!is_null(mda_get($opts,'y')) || !is_null(mda_get($opts,'yes'))) define('ANSWER_YES',true);

//figure out our target and mirror
target($opts); //sets the constant 'TARGET'
cache(); //sets the constant 'CACHE'

//init target def
TgtDef::init(TARGET);

//figure out what we are going to do
foreach(array_keys($opts) as $act){
	switch($act){
		//help
		case 'help':
		case 'h':
			usage();
			exit;
			break;

		//db actions
		case 'build-db':
		case 'b':
			$noerror=true;
			buildDb();
			break;
		case 'export-db':
		case 'e':
			$noerror=true;
			exportDb(mda_get($opts,'mirror'));
			break;
		case 'show-db':
		case 'S':
			$noerror=true;
			showDb($opts);
			break;
		case 'update':
		case 'U':
			$noerror=true;
			update();
			break;

		//package actions
		case 'upgrade':
		case 'u':
			upgrade();
			exit;
			break;
		case 'search':
		case 's':
			search(mda_get($opts,'search') ? mda_get($opts,'search') : mda_get($opts,'s'));
			exit;
			break;
		case 'install':
		case 'i':
			install(mda_get($opts,'install') ? mda_get($opts,'install') : mda_get($opts,'i'));
			exit;
			break;
		case 'remove':
		case 'r':
			remove((mda_get($opts,'remove') ? mda_get($opts,'remove') : mda_get($opts,'r')));
			exit;
			break;
		case 'purge':
		case 'p':
			remove((mda_get($opts,'purge') ? mda_get($opts,'purge') : mda_get($opts,'p')),true);
			exit;
			break;

		//backup/restore
		case 'backup':
		case 'B':
			backup(mda_get($opts,'backup') ? mda_get($opts,'backup') : mda_get($opts,'B'),mda_get($opts,'db-dump'));
			exit;
			break;
		case 'restore':
		case 'R':
			restore(mda_get($opts,'restore') ? mda_get($opts,'restore') : mda_get($opts,'R'),mda_get($opts,'db-restore'),mda_get($opts,'restore-file'));
			exit;
			break;

		//migrate
		case 'migrate':
			migrate(mda_get($opts,'migrate'));
			exit;
			break;

		//local package actions
		case 'list':
		case 'l':
			listInstalled();
			exit;
			break;

		//misc/utility
		case 'int-version':
			intVersion(mda_get($opts,'int-version'));
			exit;
			break;
		case 'clear-cache':
			clearCache();
			exit;
			break;
		case 'version':
		case 'V':
			UI::out('Tools Version: '.LSSTOOLS_VERSION."\n");
			UI::out('Admin Version: '.LSSTOOLS_VERSION."\n");
			exit;
			break;
			
		//pkg management
		case 'create':
			createPackage(mda_get($opts,'create'));
			exit;
			break;
			
		case 'delete':
			deletePackage(mda_get($opts,'delete'));
			exit;
			break;
			
		case 'export':
			exportPackage(mda_get($opts,'export'),mda_get($opts,'mirror'));
			exit;
			break;
			
		case 'refactor':
			refactorPackage(mda_get($opts,'refactor'),mda_get($opts,'dir'));
			exit;
			break;
			
		//def management
		case 'show':
			$def = getDef(mda_get($opts,'show'));
			showDef($def);
			exit;
			break;
		case 'set':
			$def = getDef(mda_get($opts,'set'));
			setValue($def,mda_get($opts,'name'),mda_get($opts,'value'));
			exit;
			break;
		case 'add':
			$def = getDef(mda_get($opts,'add'));
			addValue($def,mda_get($opts,'name'),mda_get($opts,'value'));
			exit;
			break;
		case 'del':
			$def = getDef(mda_get($opts,'del'));
			delValue($def,mda_get($opts,'name'),mda_get($opts,'value'));
			exit;
			break;

		//default
		default:
			continue;
			break;
	}
}

//no action was taken, error out
if($noerror === false) throw new Exception('No action supplied see --help for details',ERR_NO_ACTION);
